### ISSUE TITLE
Feature suggestion: Moon phase/rise/set

### ISSUE THREAD

**Issue by arschficken** [forrestguice/SuntimesWidget#52]
Your app is handy for people who interact with nature (hiking, stargazing, aurora hunting, etc.). The moon also helps or hinders these activities. But there is currently no satisfactory moon tracking app in F-Droid that would give [basic info like timeanddate.com](https://www.timeanddate.com/astronomy/usa/phoenix).

If there's a Java library for moon phase/rise/set, you may want to add this functionality to Suntimes, or create a Moontimes app.

**Comment by forrestguice** [comment]
I like this - this feature is definitely inline with the goals of the app. ...but I'm not sure if or when this will be developed. 

An issue already exists on the sunrisesunsetlib-java project for this feature: https://github.com/mikereedell/sunrisesunsetlib-java/issues/13 .. if implemented there I will add it to the app asap.

**Comment by JonnyTech** [comment]
Can you use any code from this old Android app (https://github.com/yancyn/MoonWidget) as it seems to do what is required?

**Comment by forrestguice** [comment]
possible, but I'm not certain. 

the next step to implementing this feature is to define an interface (or extend the existing one; https://github.com/forrestguice/SuntimesWidget/blob/master/app/src/main/java/com/forrestguice/suntimeswidget/calculator/SuntimesCalculator.java) with the necessary methods, and then provide an implementation of that interface (pretty much just a "wrapper" around some library), and then develop all the UI stuff against the interface - I don't want to author or maintain the libraries themselves (worrying about getting that stuff right seems like a drag tbh).

the app currently wraps two libraries (sunrisesunsetlib and recently added ca.rmen:lib-sunrise-sunset) - if either of them provided the necessary moon functionality it wouldn't be hard to extend the existing interface and get going on the UI stuff immediately. otherwise I'm gonna have to figure out what those methods need to be on my own (teach myself a bunch of stuff), find a good implementation, etc. definitely more work. ... the linked code looks interesting, might be useful to study, but what we really need is an active project that can shoulder the burden of getting it right (if that makes any sense).

**Comment by MenoData** [comment]
The suggested widget from github.com/yancyn seems to be outdated (last commit in year 2013) and only offers moon phases until year 2016 (see [this file](https://github.com/yancyn/MoonWidget/blob/master/res/xml/moon.xml) which forms a small list of dates for new moon).

But another suggestion to finish this issue:

I have also tackled the underlying problem in my lib Time4J (resp. the Android variant [Time4A](https://github.com/MenoData/Time4A)) and can at least offer moon phases (and solstices, sunrise, sunset and solar noon with higher precision, too). The remaining issue would just be moon rise/set which [I want to tackle](https://github.com/MenoData/Time4J/issues/704) still within this year 2017.

**Comment by forrestguice** [comment]
sweet. Your library looks very nice. Thanks for letting me know its available. I've been taking a pretty long break from this project, but now I'm thinking I will resume. The solstice related stuff is next, but I can now see a moon widget on the horizon. :)

**Comment by MenoData** [comment]
If it is okay for you then I can also submit a PR to implement your `SuntimesCalculator` as base for further refinements. 

There is also another thing to consider: **Which timezone data should be used by Time4A**? Remember that every version of Time4A has its own tz-data. But it is possible to configure Time4A such that it uses the platform tz-data of the mobile phone instead. If we go with the timezone data of Time4A then your suntimes-widget would ideally need to be updated with every new tz-version resp. Time4A-version. If we go with the platform tz-data instead then you have less update-work, but the burden is on the mobile phone user (I assume that most users would just not update their OS and maybe change their system tz to another one with suitable tz-offset in order to correct for wrong local times).

**Comment by forrestguice** [comment]
Thanks, I really appreciate your help with that. I'm working in branch time4a (https://github.com/forrestguice/SuntimesWidget/tree/time4a) on this file..

https://github.com/forrestguice/SuntimesWidget/blob/time4a/app/src/main/java/com/forrestguice/suntimeswidget/calculator/time4a/SunriseSunsetSuntimesCalculator.java 

As far as timezones go, I'd like to continue to use the platform data (for time being at least).

I think a trade off in shipping the latest timezones is an increase in apk size? I'd like to minimize that (but do expect a hefty increase in size all the same). Otherwise, I think shipping an up-to-date list of timezones is a very nice feature, especially for older devices (which I want to continue supporting).

**Comment by MenoData** [comment]
The new release of Time4A (3.38-2017c) has now support for moonrise/moonset, too.

**Comment by MenoData** [comment]
I am looking at your first coding in the moontime-branch and see:

233     /** 
234      * Moonrise 
235      * @param date a Calendar representing a given date 
236      * @return a Calendar for moonrise for the given date 
237      * @since 1.3.0 
238      */ 
239     Calendar getMoonriseCalendarForDate(Calendar date); 
240 
 
241     /** 
242      * Moonset 
243      * @param date a Calendar representing a given date 
244      * @return a Calendar for moonset for the given date 
245      * @since 1.3.0 
246      */ 
247     Calendar getMoonsetCalendarForDate(Calendar date); 

My suggestion: It would be better for sake of performance to only have one method returning maybe an array consisting of moonrise/moonset or a custom pair class. Keep in mind that the computations of moon events is more consuming than sun calculations so calling one method is better than calling two different methods twice.

If you look at my API then you will see a similar approach, namely calling only once on an instance of `LunarTime` obtaining an instance of [LunarTime.Moonlight](http://time4j.net/javadoc-en/net/time4j/calendar/astro/LunarTime.Moonlight.html) which finally contains informations about moonrise AND moonset at the same time.

**Comment by forrestguice** [comment]
good catch. There really isn't a situation where only one method needs to be called (vs both).  I wasn't thinking about performance while writing those (just mimicking the existing patten). imho a pair class is better style, but I think returning a Calendar[] should suffice.

**Comment by forrestguice** [comment]
thinking about this a bit more.. The design of the interface (up to this point) is synchronous/blocking and is run directly on the UI thread in some cases. While I've noticed performance variation between the different solar calculators while logging (the 'simple' one of course the fastest), those didn't seem costly enough to warrant a major change. 

It could be rewritten to be async in nature - methods changed to void and instead using listeners/callbacks to return results. Supposing I wanted to allow implementations to make network calls (to some third party service that does the calculations) I think this would be required. ..but allowing (or ever needing) network is contrary to app's philosophy (so I haven't bothered making it possible).

I could also just wrap it all up in an AsyncTask if there is real potential to block the UI thread. Are any of these methods that costly? or should the current approach work? (as long as we're mindful to optimize and avoid unnecessary work).

**Comment by MenoData** [comment]
I don't know if the performance is a big problem now. Suggest to simply test it on a slow device. Of cource, I could also release an extra variant of moon calculations in lower accuracy to speed it up (but then deviations of 4-5 minutes or more are to be expected even for normal geographic latitudes).

**Comment by forrestguice** [comment]
I understand. I don't think performance will be an issue but ready to address it need be.

I'd wait until it tests poorly (on some api10 device) before adding 'simple' moon calculations. 4-5 minutes error does sound like a lot (although probably still accurate enough for most uses).

**Comment by forrestguice** [comment]
I will tag v0.7.0 this weekend. With any luck it will be available in F-Droid by next week. This release adds basic widgets - integration w/ the app and additional widgets are still to come. 

![moonwidget4_1x1_preview](https://user-images.githubusercontent.com/10246147/36186737-209a31de-10fe-11e8-90e7-6ed1dc12afe4.png)
![moonwidget0_1x1_preview](https://user-images.githubusercontent.com/10246147/36186738-20b5baa8-10fe-11e8-8ea8-357f782f71bc.png)
![moonwidget0_2x1_preview](https://user-images.githubusercontent.com/10246147/36186739-20d24c86-10fe-11e8-9e8d-c8ddc68140bf.png)
![moonwidget0_3x1_preview](https://user-images.githubusercontent.com/10246147/36186740-20ff6cf2-10fe-11e8-9b3d-7f5e236a560a.png)

![moonwidget3_1x1_preview](https://user-images.githubusercontent.com/10246147/36186736-207e1bc0-10fe-11e8-8138-6b22f2cef834.png)
![moonwidget2_1x1_preview](https://user-images.githubusercontent.com/10246147/36186735-20663d34-10fe-11e8-9f5c-8338970d8d13.png)
![moonwidget1_1x1_preview](https://user-images.githubusercontent.com/10246147/36186741-211b82b6-10fe-11e8-9211-e430a9542834.png)

**Comment by JonnyTech** [comment]
Nice. Maybe it is time to change the name of the app...? ;)

**Comment by forrestguice** [comment]
Ok, I think we are there (so to speak), or nearly there. The moon functionality is now pretty well integrated w/ v0.8.0, so I'm going to close this issue.

I can actually think of a few more moon related enhancements. For example, my calendar is missing the major moon phase dates - I think that is something the app can provide through a ContentProvider. 

If you think of anything, please don't hesitate to open another issue.

**Comment by forrestguice** [comment]
@JonnyTech lol. I've tried really hard not to think about it. ;) I'm not sure what I would rename it to, and not sure the ideas I've had are any better (more descriptive). ..but "Suntimes" could just as well be a newspaper :\

maybe this app is more of a "Sundog". I really dunno.

**Comment by arschficken** [comment]
Thanks! Saw the update come through yesterday and I'm really amped about it. Being able to drill down to more info by tapping on the Moon section is great US design too.

By the way, what does the single time and bearing in the top right hand corner of that Moon dialog signify?

**Comment by forrestguice** [comment]
That rightmost time in the moon dialog is tomorrows rising or setting time (whichever occurs first). Supposing the moon sets first, then rises, it will show setting time tomorrow. Similar, if the moon rises then sets, it will show rising time tomorrow.

Its a bit superfluous considering the today/tomorrow cards show the exact same thing. I was also considering making this a 2x1 moon widget (where is may actually be a lot more useful).
